### ISSUE TITLE
T450s Wifi replacement

### ISSUE THREAD

**Issue by Coonaps** [jsassu20/Lenovo-ThinkPad-T450-Mojave#23]
Does the T450s (smaller body) have enough space to replace the Intel Wifi with an Airport Card + NGFF Adapter as recommendend in your guide? Any first hand experience?

**Comment by jsassu20** [comment]
I’m not sure, it’s possible that it might have the same clearance but I don’t have one so I can’t be sure. However there are other adapters which can be used. You could just purchase the adapter and modify if by cutting off the back plate and using it just for the adapter portion

**Comment by willmav5000** [comment]
It's a very tight fit and you'll either have to skip tightening the screw closest to the adapter or find another wifi card that works without needing an adapter.

I'm currently using the Dell DW1820A (BCM94350ZAE), it can be a little tempermentle and I get the occational kernel panic on boot but other than that it's been working great with bluetooth and airdrop support.

**Comment by i3p9** [comment]
@willmav5000 are you using the dell card on your T450s? Does handoff/airdrop works for you?

**Comment by willmav5000** [comment]
@i3p9 handoff/airdrop is working fine and this is the card I'm using https://www.ebay.co.uk/itm/Dell-DW1820A-BCM94350ZAE-5G-Dual-Band-867Mbps-M-2-NGFF-Wireless-WiFi-Card-W042/303110498999

It took a little playing around to get it working after upgrading to Catalina but it's been great so far.

**Comment by i3p9** [comment]
@willmav5000 wow that's significantly cheaper than one of the aliexpress option I was looking at. Even with shipping from UK to US it's going to cheaper. Thank you! What did you do for Catalina support? 

I'm gonna attempt to hackintosh my T450s after the card arrives, would be really helpful if you could help if I get stuck on something, if that's not too much? Do you have Discord/Email to contact?

**Comment by willmav5000** [comment]
@i3p9 yeah the issue with these cards is apparently they will either work or they won't but as they're so cheap I thought I give it a try anyway before buying one of the more expensive ones. You could probably get one for around the same price from ebay.com so it gets delivered quicker.

The upgrade to catalina broke a few things for me like the wifi, bluetooth and sound. There were a number of things I did like updating the kexts, created a modifeid version of IO80211Family.kext and used Clover to patch the card id to a supported one.

I'm not exactly sure which one(s) got the card working but I'll put together a file with the config and kexts I'm using and help you out where I can.

Most of the things I tried were because people said they'd had some success with over at https://osxlatitude.com/forums/topic/11540-dw1820a-the-general-troubleshooting-thread/

**Comment by i3p9** [comment]
@willmav5000 thank you so much! And I’ve already ordered the card, ebay.com options weren’t very good to be honest. Mostly shipped from China. 

And yeah I’ve also been researching the card and seems like custom versions of kexts are working for the card. 

And yeah having your EFI with acpi patches and everything would definitely help me getting a starting point. Since we have the same model. 

Do you have the i5 or the i7 version? 

Thank you so much!

**Comment by i3p9** [comment]
@willmav5000 Sorry if I'm being annoying but do share the T450s EFI you're using whenever you get some free time! Thanks.

**Comment by willmav5000** [comment]
@i3p9 sorry about that, I've been quite busy recently and it just totally slipped my mind. Here's a link to my EFI folder and the modified IO80211Family.kext I'm using https://mega.nz/#F!6zJXhSAD!M4qMhml0JMAOWmaCObFyLA

You'll have to go over the clover config as I've added some stuff in there specific to my laptop to fix the available RAM, etc. It should mostly be the same as the one from this repo but I'm not sure what else I've done.

**Comment by i3p9** [comment]
@willmav5000 Thank you! It's not a problem at all. I haven't started the process yet. 

As for the RAM I've got 12GB installed as it's the max T450s supports. Thing is I wanna make sure I understand what you did in the config and whatnot, so I might have some questions so I can fix stuff later, or troubleshoot properly. 

I opted to go with OpenCore, but I'll take notes from your EFI for sure. 

Can you let me know which changes you made for the Wifi card?

**Comment by willmav5000** [comment]
@i3p9 I tried to setup OpenCore but couldn't get it to work.

I've got 20GB RAM installed and it's correctly detected in the BIOS and works fine.

I'll compare what I've got against the config from the repo and try to clean it up a little then let you know what's actually needed.

If you have a working OpenCore config would you mind sharing it? I'd love to know what I missed.

**Comment by willmav5000** [comment]
@i3p9 well this should be easy enough for you, I reverted back to the config from the repo and it looks like only thing you need to do is add brcmfx-driver=1 to your boot arguments while using the modified IO80211Family.kext file.

**Comment by i3p9** [comment]
@willmav5000 So just the boot args and  copy the kext to System? Gotcha. And you kept all the other kexts as well right? 

I actually started my OpenCore Config from this repo, he's pretty great: https://github.com/EchoEsprit/Hackintosh-Catalina-OpenCore-Lenovo-T450s-efi

Now I'm trying to get the Card Reader working, only thing that's important to me that isn't working. I have a lot of reading left to do on ACPI patches, as I'm a newcomer to laptop hackintosh.

**Comment by willmav5000** [comment]
@i3p9 I'm not sure if you'll need all the other kext files, but one that you will need for the wifi is AirportBrcmFixup.kext as I've just tried it with that OpenCore repo.

I'm not sure which of the others enables bluetooth but I'll let you know if I find out.

**Comment by i3p9** [comment]
@willmav5000 thank you so much! 

And another question, what’s your BIOS version? I have the latest one and it doesn’t have wifi whitelist right?

**Comment by willmav5000** [comment]
@i3p9 yeah Lenovo removed the whitelist a while ago and I'm on the latest BIOS too.

**Comment by willmav5000** [comment]
@i3p9 I have bluetooth working now too, you'll need to add BlueTooth_BCM.kext, BrcmBluetoothInjector.kext, BrcmFirmwareData.kext and BrcmPatchRAM2.kext to the config.

![image](https://user-images.githubusercontent.com/272902/67531933-72187200-f6bc-11e9-8116-9f6871b72767.png)

**Comment by i3p9** [comment]
@willmav5000 oh wait you have dw1820 working in Opencore as well? That’s great news! 

So you added AirportBrcmFixup.kext, BlueTooth_BCM.kext, BrcmBluetoothInjector.kext, BrcmFirmwareData.kext and BrcmPatchRAM2.kext for the wifi and bt?

Did you also put the modified IO80211Family.kext in system as well? Plus the boot args?

**Comment by willmav5000** [comment]
@i3p9 yeah WiFi, bluetooth, AirDrop and Handoff are all working on OpenCore with those files and the brcmfx-driver=1 boot argument.

I put AirportBrcmFixup.kext, BlueTooth_BCM.kext, BrcmBluetoothInjector.kext, BrcmFirmwareData.kext and BrcmPatchRAM2.kext in the OpenCore Kexts folder then added them to the config. 

Then put the modified IO80211Family.kext file in /Library/Extensions/ and ran this "sudo chown -R root:wheel /Library/Extensions; sudo chmod -R 755 /Library/Extensions; sudo kextcache -i /".

**Comment by i3p9** [comment]
@willmav5000 Cheers! You've been super helpful in helping me understand. I'll get back to you with probably more questions (haha) when I get Airport card. I ordered both, the one you're using, and the one from the OpenCore guide. If I get the correct serial dw1820 I'll return the other. Since there's  multiple version of the card and I don't wanna wait again. ;)

**Comment by willmav5000** [comment]
@i3p9 no problem, I'm gonna have a play around with OpenCore a little as the trackpad is a little sketchy and the volume is a little too low.

The DW1820 I got is supposed to be one of the ones that doesn't work or has issues and the MAC address didn't match the label but I tried it anyway and it's been working great.

**Comment by i3p9** [comment]
@willmav5000 Maybe the VoodooPS2Controller is different in the OC one? Not sure. And please let me know if you fix those things. I'd rather use OpenCore than Clover. OC is easier to manage, understand an structured to be honest haha. 

And yeah DW1820 has multiple versions. osxlatitude has a thread about it but the site is unreachable right now. I'll send you the link later. :)

**Comment by i3p9** [comment]
@willmav5000 did you figure out what's wrong with the touchpad on the OC setup? I like this repo's touchpad but OC one is bad.

**Comment by willmav5000** [comment]
@i3p9 the latest commit on the OC repo has fixed the touchpad issues for me. I've also got the card reader working so I've just created a pull request to have it merged.

**Comment by jsassu20** [comment]
Be careful with the DW 1820a There’s four revisions of that card one comes in Del one comes in Lenovo anyway only one of the four revisions works without issues the other two don’t work at all from what I can tell and one of them causes Colonel panics I actually went from the DW 1820a to An actual airport card and after doing that I had no problems I recommend doing the airport card because you’re going to get perfect Mac functionality out of it without any necessary adjustments it’ll work out of the box

**Comment by jsassu20** [comment]
I still have my DW1820a If one he wants it just send me like 10 bucks and I’ll mail it to you just use my PayPal

**Comment by jsassu20** [comment]
![image](https://user-images.githubusercontent.com/12767985/67625831-43f87680-f811-11e9-9690-545723ae45ff.jpg)
![image](https://user-images.githubusercontent.com/12767985/67625837-5bcffa80-f811-11e9-9d8e-9d04204746f7.jpg)

**Comment by willmav5000** [comment]
@jsassu20 I tried the airport card with the adapter first but it's a tight fit for the t450s and I ended up snapping one of the screw sockets while screwing the bottom back on.

The 1820a I ended up with isn't the Dell or Lenovo one, it's actually one of the ones that isn't supposed to work (1028:0023) and it originally kept giving me kernel panics, making my machine reboot or stalling while booting.

I've had almost zero issues with the card since using a modified IO80211Family kext which is just a copy of the system one with AirPortBrcmNIC and IO80211NetBooter removed and pci14e4,43a3 added to IONameMatch in the Info.plist for AirPortBrcm4360.

**Comment by i3p9** [comment]
@willmav5000 Yup the touchpad issue is sorta fixed for me. Two finger scrolling is very slow for me, how is it for you? I tried setting it to the highest in accessibility settings but no dice. :(

**Comment by fabPress33** [comment]
Hi! I own T450 with i5-5300u with 8gb of ram and ssd and a crad wifi BCM94350ZAE ..... can anyone help me make my hackintosh? give me some advice etc .... I am a graphic designer and I would love to have a spare notebook ...

**Comment by caifenglin** [comment]
> @jsassu20 I tried the airport card with the adapter first but it's a tight fit for the t450s and I ended up snapping one of the screw sockets while screwing the bottom back on.
> 
> The 1820a I ended up with isn't the Dell or Lenovo one, it's actually one of the ones that isn't supposed to work (1028:0023) and it originally kept giving me kernel panics, making my machine reboot or stalling while booting.
> 
> I've had almost zero issues with the card since using a modified IO80211Family kext which is just a copy of the system one with AirPortBrcmNIC and IO80211NetBooter removed and pci14e4,43a3 added to IONameMatch in the Info.plist for AirPortBrcm4360.

I also use 1820a now, but on my t450s, WiFi can't work well. Can you help me

**Comment by costadaniel** [comment]
@jsassu20 you can show the ngff adapter and post the link for buying one?

**Comment by jsassu20** [comment]
Sure thing when I get home I actually just purchased another adapter and bcm94360cs2 card for another dell computer I just hought on eBay. I got the adapter for 8 bucks and the card for 5 bucks. I would def recommend eBay now since the first time I paid 50 total for both and only 13 this time. Both pieces came in less than a week

**Comment by jsassu20** [comment]
![CF02993D-594B-4F55-8F67-5E41769C1D7E](https://user-images.githubusercontent.com/12767985/70750074-a7743000-1cfb-11ea-94a8-2e4570b824a4.png)

**Comment by costadaniel** [comment]
I purchased the card and adapter in Aliexpress some days ago, i have to unlock or make something in bios to use the card?

**Comment by jsassu20** [comment]
> > @jsassu20 I tried the airport card with the adapter first but it's a tight fit for the t450s and I ended up snapping one of the screw sockets while screwing the bottom back on.
> > The 1820a I ended up with isn't the Dell or Lenovo one, it's actually one of the ones that isn't supposed to work (1028:0023) and it originally kept giving me kernel panics, making my machine reboot or stalling while booting.
> > I've had almost zero issues with the card since using a modified IO80211Family kext which is just a copy of the system one with AirPortBrcmNIC and IO80211NetBooter removed and pci14e4,43a3 added to IONameMatch in the Info.plist for AirPortBrcm4360.
> 
> I also use 1820a now, but on my t450s, WiFi can't work well. Can you help me

Why don't you try using a ngff a/e extension cable its two boards with a ribbon between them one plugs into the wifi slot as a single card and the airport cad would fit into the end of the adapter instead of on top of it like I have. You'll have to secure the card in another area but you should be able to find room for it like this.

https://www.amazon.com/NGFF-Mini-PCI-Adapter-Cable/dp/B07JFYSNVL/ref=pd_sbs_147_2/133-1039478-9926862?_encoding=UTF8&pd_rd_i=B07JFYSNVL&pd_rd_r=88de14da-7aec-44e5-9d2e-ca523e13e8ff&pd_rd_w=vw1o6&pd_rd_wg=CUHLo&pf_rd_p=5873ae95-9063-4a23-9b7e-eafa738c2269&pf_rd_r=DTYCTPJGY7H5BKB24664&psc=1&refRID=DTYCTPJGY7H5BKB24664

**Comment by jsassu20** [comment]
![513FAuqmhxL _SL1000_](https://user-images.githubusercontent.com/12767985/70842418-74ac6380-1df1-11ea-8724-944b3c45f3fd.jpg)

also with the adapters you can cut as low as the yellow line same with the one you snapped, its just support material above the yellow line

**Comment by rahmadsandy** [comment]
Hello, thanks for @jsassu20 for provide this great EFI Repo, i have modified config list from @willmav5000 ,
all features almost works perfectly , even my DW1802A (BCM94350ZAE) , wifi and bluetooth works
but i have little problem with it, after a while about 20 minutes i used my laptop suddenly cursor running slowly than stuct can do anything, so i hard power off , and its happend continues manner,
can you help how to debuging this case?  so i can track whats happend?

sorry for my english

**Comment by jsassu20** [comment]
you have to tape the pins for the power switch that's present on pc but not in Mac

**Comment by jsassu20** [comment]
https://osxlatitude.com/forums/topic/11540-dw1820a-the-general-troubleshooting-thread/?do=findComment&comment=91179

**Comment by truonghdpk** [comment]
I have this version in my country at Vietnam DW1820A - 096JNT, i don't know does it work correctly ? @jsassu20 
![Screen Shot 2020-02-07 at 12 32 41](https://user-images.githubusercontent.com/45913828/74003646-f067fe00-49a5-11ea-872d-3e5ceece6393.png)

**Comment by jsassu20** [comment]
Okay you're all set boss all DW1820a cards are now fully supported the issue surrounding the cards has to do with ASPM being enabled which causes instability in 4 out the 5 different variations of this card. Disable asp and all of them function properly 100% here's a link to the info you need 

https://osxlatitude.com/forums/topic/11322-broadcom-bcm4350-cards-under-high-sierramojavecatalina/?do=findComment&comment=96996

**Comment by jsassu20** [comment]
> Hello, thanks for @jsassu20 for provide this great EFI Repo, i have modified config list from @willmav5000 ,
> all features almost works perfectly , even my DW1802A (BCM94350ZAE) , wifi and bluetooth works
> but i have little problem with it, after a while about 20 minutes i used my laptop suddenly cursor running slowly than stuct can do anything, so i hard power off , and its happend continues manner,
> can you help how to debuging this case? so i can track whats happend?
> 
> sorry for my english

refer to my previous post sorry for late response!

**Comment by jsassu20** [comment]
> I have this version in my country DW1820A - 096JNT, i don't know does it work correctly ? @jsassu20
> ![Screen Shot 2020-02-07 at 12 32 41](https://user-images.githubusercontent.com/45913828/74003646-f067fe00-49a5-11ea-872d-3e5ceece6393.png)

refer to my previous comment. Should be all set!

**Comment by jsassu20** [comment]
in clover properties for DW1820a just need to ad property key: pci-aspm-default with the value: 0 as a Number

**Comment by jsassu20** [comment]
![ASPM_injection jpg 2ec33e393b69e45712e8b321963ecbfc](https://user-images.githubusercontent.com/12767985/74301185-3dd6c900-4d20-11ea-87cd-71e2f724e2c9.jpg)

**Comment by jsassu20** [comment]
All credit goes to Herve (https://osxlatitude.com/profile/4953-hervé/) and the people who posted this info in his topic area.

**Comment by truonghdpk** [comment]
@jsassu20 
![84652230_2316859288618209_8506457577570172928_n](https://user-images.githubusercontent.com/45913828/74453521-17966200-4eb5-11ea-8cd6-8562dfd8f2d9.jpg)
i get this card (DW1820A - 096JNT) in my laptop.i'm trying but before and after i update your config @jsassu20 my laptop still get stuck in apple logo anyway i can't, i must go to BIOS to disable wifi so it pass, i'm sure my card it fine with no error i'm trying it in Window and it work correctly, why ?

**Comment by jsassu20** [comment]
then your doing something wrong because the patch in not loading. Check your IOREG to ensure your names match

**Comment by truonghdpk** [comment]
I'm using Thinkpad T450s and i using your kext from this repo of course i just paste your EFI file and rename config_DW1820a_Working.plist to config.plist

**Comment by truonghdpk** [comment]
> then your doing something wrong because the patch in not loading. Check your IOREG to ensure your names match
> 
> have you taped the pins? What version of Mac are you on and where are your kext files loading from?
@jsassu20 
I'm using Thinkpad T450s and i using your kext from this repo of course i just paste your EFI file and rename config_DW1820a_Working.plist to config.plist, my macversion is catalina

**Comment by truonghdpk** [comment]
@jsassu20  like this https://www.tonymacx86.com/threads/install-dell-dw1820a-broadcom-bcm94350zae.265130/page-11

**Comment by jsassu20** [comment]
Boot verbose and paste the screen contents when the text stops

**Comment by rogerters** [comment]
First of all, thanks for the incredible work you have put into this project. I can only guess how much time and work you have put into this. My t450s works literally without any flaws except for this wifi issue.
I have the exact same issue as truonghdpk and I just don't get what's the problem. The issue is also that I cannot figure out the correct PciRoot path as I cannot even boot into my macOS install with the DW1820a activated in BIOS.
Verbose mode doesn't help either as the screen freezes after it switches to the last part of the boot process when the apple logo pops up.
Would really appreciate any help you can provide her :)

**Comment by rogerters** [comment]
update: I did get wifi functionality to work by ticking the fixairport box in clover. Bluetooth also seems to work with the help of BrcmPatchRAM2.kext, BrcmFirmwareData.kext and BrcmBluetoothInjector.kext. However, what still has not been working is handoff e.g.
Airdrop, however, works... really don't know what's going on her. Furthermore, 5g wifi just does not work at all. I cannot find any networks...

**Comment by jsassu20** [comment]
> First of all, thanks for the incredible work you have put into this project. I can only guess how much time and work you have put into this. My t450s works literally without any flaws except for this wifi issue.
> I have the exact same issue as truonghdpk and I just don't get what's the problem. The issue is also that I cannot figure out the correct PciRoot path as I cannot even boot into my macOS install with the DW1820a activated in BIOS.
> Verbose mode doesn't help either as the screen freezes after it switches to the last part of the boot process when the apple logo pops up.
> Would really appreciate any help you can provide her :)
> 
> EDIT: let me add that I am running a Mojave installation

PciRoot(0x0)/Pci(0x1c,0x1)/Pci(0x0,0x0) is the root path

the installer fails to load because the dw1820a shares a common Device ID with a kernel extension that is not supported by the AirPortBrcmNIC.kext which that device ID attempts to load. Disable the wifi card and then boot into Mac and find the IO80211Family.kext in /S/L/E and delete the AirPortBrcmNIC kext from inside of it and the system will boot with the card. Or add the "pci-aspm-default" property to all of your pcie device properties

**Comment by jsassu20** [comment]
or try this config first 

[config_DW1820a_Working.plist.zip](https://github.com/jsassu20/Lenovo-ThinkPad-T450-macOS-Catalina/files/4567737/config_DW1820a_Working.plist.zip)

**Comment by rogerters** [comment]
@jsassu20 thank you for your answer. deleted AirPortBrcmNIC inside of IO80211Family.kext from S/L/E and repaired the cache. Then I used your config_DW1820a_Working.plist in combination in combination. Still no luck. When I tick the FixAirport box it boots again and everything works except for laggy bluetooth and no 5G Wifi

**Comment by jsassu20** [comment]
oh sorry change the compatible property to pci14e4,43a0

**Comment by rogerters** [comment]
@jsassu20 I tried that as well based on your config. But still hangs when it switches to the apple logo. So, same issue as before. Really don't know what I'm doing wrong here. I can only boot if I tick "fixairport". But then I do not have any 5G networks available. Do u know what fixairport actually does? as I could not find any information on that online

**Comment by gabgabe** [comment]
Hello guys,

Absolutely had problems fitting the card. The cart itself worked for a few monts, I tried to fit it as best possible, with spacer, even made an aluminium foil flat heatsink to put on the top of the card, thermal paste under the card, stretch bottom cover with head gun... nope, the card was parallel and followed all other people said in this issue but sometimes if I move the laptop the card disconnects so I can’t use my thinkpad as usually. My pc always fall one time a month so I need to fix this card problem.

And here we gooooo, look at the dirty job I made, works perfectly now ;)

Byeee
![6623D463-C645-416E-A586-481CDC23663E](https://user-images.githubusercontent.com/45693996/83782124-ad861080-a68f-11ea-98db-389609589042.jpeg)
![7D6BFFE1-208B-406F-B326-9C81B430EAA3](https://user-images.githubusercontent.com/45693996/83782135-af4fd400-a68f-11ea-86ea-244fdcd4eed7.jpeg)
![88015399-8AE2-436D-958E-25A296264CCA](https://user-images.githubusercontent.com/45693996/83782138-af4fd400-a68f-11ea-863e-47398156f4e9.jpeg)
![62E1C05A-B2ED-434A-A28F-B65D035EA8D4](https://user-images.githubusercontent.com/45693996/83782140-afe86a80-a68f-11ea-87a2-f9587086fbd7.jpeg)
![B3E33264-0D90-478A-A6E2-F9CBECBED256](https://user-images.githubusercontent.com/45693996/83782143-b1199780-a68f-11ea-90bd-8f2932f1e80c.jpeg)
![DFCA67FB-4F77-4FDE-BEC6-C1E03790A90F](https://user-images.githubusercontent.com/45693996/83782147-b1b22e00-a68f-11ea-83ce-b83d58c60560.jpeg)
![86952535-5910-4BEC-82CF-580C6A7AA06E](https://user-images.githubusercontent.com/45693996/83782149-b24ac480-a68f-11ea-93ba-f6eb2c1dfd03.jpeg)
![BAA7C71C-BC71-4483-BF3D-24213622809F](https://user-images.githubusercontent.com/45693996/83782151-b2e35b00-a68f-11ea-93d1-11a31160cbd0.jpeg)

**Comment by svobluk** [comment]
You can use similar to this. This gives you also possibility to move card out of the original location.

https://a.aliexpress.com/_dSJdQ54
